### PR TITLE
My resources page optimization

### PR DESCRIPTION
**Pull request by pkdash**
### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [ ] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [ ] Peer Code review and approval

### Positive Test Case
1. In order to see a noticeable improvement in page loading time for my resources page, You need to have at least 50 resources. If you have 50 or more resources locally, make a note of page loading time for my resource page with your current working branch. Also make a note of loading time for a specific resource landing page. Then switch to branch '3211-my-resources-optimization' branch. Again load my resources page and compare the page load time with the one you noted earlier. You should see at least 30% improvement in page loading time. Use the same resource for landing page loading time comparison and you should see some improvement here too.

### CODE REVIEW

**Comment by pkdash** [comment]
@cuahsimarko, @sblack-usu and @Maurier I have created this PR to get some initial feedback on the changes I have made to improve page load time.

**Comment by cuahsimarko** [comment]
Awesome, I will try to get time to review early this week.

File: hs_core/views/utils.py [hunk 1]
```diff
@@ -536,7 +537,8 @@ def create_form(formclass, request):
     return params
 
 
-def get_my_resources_list(request):
+# TODO : delete this function
+def get_my_resources_list_old(request):
```
Comment (cuahsimarko) [inline hs_core/views/utils.py]: Are you leaving this for our sake as a part of the review?

File: hs_core/templates/pages/my-resources.html [hunk 2]
```diff
@@ -262,11 +262,13 @@ <h4 class="panel-title">
                             {# Last Modified #}
                             <td>{{ res.last_updated|date:"M d, Y" }} at {{ res.last_updated|time }}</td>
                             <td>
+                                {# Here we are hitting the database once for each resource that is listed #}
```
Comment (sblack-usu) [inline hs_core/templates/pages/my-resources.html]: I wonder if we could use prefetch to speed these loops up.

File: hs_core/views/utils.py (hunk 1 above)
Comment (pkdash) [reply hs_core/views/utils.py]: I left it for my testing to see the difference between the old query and the new one. In my final PR I will delete it.

**Comment by sblack-usu** [comment]
@pkdash - do you have any metrics for how much this improves the response time?

File: hs_core/templates/pages/my-resources.html (hunk 2 above)
Comment (pkdash) [reply hs_core/templates/pages/my-resources.html]: Since 'metadata' is a property and not a model field, won't be possible to prefetch.

**Comment by pkdash** [comment]
@sblack-usu I don't have exact numbers for the improvement in response time yet. I was mostly looking at the browser response time as I was modifying the code for optimization.

**Comment by cuahsimarko** [comment]
@pkdash Some marginal improvements from what I can observe locally, plus it's just cleaner code now. I'll let Scott approve - looks good.

**Comment by pkdash** [comment]
It seems DB queries are not that expensive compared to the number of records that the template engine need to process. For 100 resources it should take around 10 sec to load. Page load time seems to be directly proportional to number of resources the template engine has to process.

**Comment by pkdash** [comment]
I will be cleaning up this PR in day or two.

**Comment by sblack-usu** [comment]
@pkdash - that is consistent with what I was finding when profiling my-resources.  One thing I was kicking around is putting the my-resources listing call into a separate response from the initial one.  Then it could feed the list of resources in chunks, allowing the my-resources page to begin rendering sooner.

**Comment by sblack-usu** [comment]
that may make it hard to handle filtering, but I haven't looked into it yet.

**Comment by cuahsimarko** [comment]
@sblack-usu Feeding it in pieces/lazy loading complicates the sort by

**Comment by sblack-usu** [comment]
@cuahsimarko... couldn't we just block filtering/sorting until the whole list is loaded?

**Comment by cuahsimarko** [comment]
@sblack-usu technically yes, the implementing engineer would obviously need to run by stakeholders. Also, the changes on back/front to get that working are the reason why I deferred work on that task for now, effort vs edge case (low/mid priority of having over 40 resources)... since 40 resources is generally ok, especially with Pabitra's changes. The original performance concerns were sparked by a total nonexistent/edge case, where 400 resources were present.

**Comment by pkdash** [comment]
If we want we can do the sorting at DB level if that would help lazy loading on front end.

File: hs_core/templates/pages/my-resources.html (hunk 2 above)
Comment (sblack-usu) [reply hs_core/templates/pages/my-resources.html]: Maybe we can remove this abstraction once we've moved away from other resource types.

File: hs_core/views/__init__.py [hunk 3]
```diff
@@ -1176,7 +1176,7 @@ def my_resources(request, page):
 
     context = {
                 'collection': collection,
-                'numOwned': len(resource_collection)
+                'numOwned': resource_collection.count()
```
Comment (sblack-usu) [inline hs_core/views/__init__.py]: is there a benefit to this, or do you just prefer using count?  I assume count is calculated in the db, vs len is via python, but haven't verified.

File: theme/templates/includes/res_type_col.html [hunk 4]
```diff
@@ -1,67 +1,67 @@
 {% load hydroshare_tags %}
-
-<span class="resource-type-text">{{ resource|resource_type }}</span>
-{% if resource|resource_type == "Generic" %}
```
Comment (sblack-usu) [inline theme/templates/includes/res_type_col.html]: what is the benefit to using . instead of | ?

Comment (pkdash) [reply theme/templates/includes/res_type_col.html]: The benefit is that we are not making a function (template filter) call rather accessing a field/attribute of the resource

File: hs_core/views/__init__.py (hunk 3 above)
Comment (pkdash) [reply hs_core/views/__init__.py]: 'resource_collection' is now a QuerySet, so can't use 'len()' unless first convert it to a list and we don't want to convert it to a list.

Comment (pkdash) [reply hs_core/views/__init__.py]: Yes 'count' is computed at the db.

**Comment by pkdash** [comment]
@cuahsimarko and @sblack-usu The changes I made last night causing some test to fail. I know why they are failing and should easily be fixed. The code changes that I made for accessing the resource 'metadata' property seems to improve page load time that is noticeable. We should be able to get 100 resources now in 6-7 secs down from 10-12 secs. This is what I got in my local dev environment. I would expect better results in production environment.

**Comment by cuahsimarko** [comment]
@pkdash understood. @sblack-usu @Maurier - Alva just gave me some great insight into what's going on here. After this PR wraps up, please reach out to me directly, if you are going to spend any more time on this issue. I want to avoid overlapping efforts.

**Comment by pkdash** [comment]
@cuahsimarko I have no plan to work on my-resources page after this PR.

**Comment by pkdash** [comment]
@sblack-usu This is now ready for another round of review. With the latest code changes in this PR, one should see noticiable improvement in my resources page loading time.

File: hs_core/models.py [hunk 5]
```diff
@@ -2076,15 +2076,12 @@ def get_metadata_xml(self, pretty_print=True, include_format_elements=True):
 
     def _get_metadata(self, metatdata_obj):
```
Comment (sblack-usu) [inline hs_core/models.py]: The method name and comment are not descriptive what's going on in this method.

Comment (sblack-usu) [reply hs_core/models.py]: Can line 2058 be moved to the else statement in this method?

File: hs_core/models.py [hunk 6]
```diff
@@ -2076,15 +2076,12 @@ def get_metadata_xml(self, pretty_print=True, include_format_elements=True):
...
-        res_type = ContentType.objects.get_for_model(self)
-        self.content_object = res_type.model_class().objects.get(id=self.id).content_object
+
         if self.content_object:
             return self.content_object
         else:
             metatdata_obj.save()
-            self.content_type = md_type
-            self.object_id = metatdata_obj.id
+            self.content_object = metatdata_obj
```
Comment (sblack-usu) [inline hs_core/models.py]: Is this lazy loading actually necessary?  content_object is a generic foreign key.  If we could set this field eagerly we could do a prefetch_related on content_object and speed things up some more.

File: hs_app_timeseries/models.py [hunk 7]
```diff
@@ -971,6 +971,7 @@ def series_names(self):
         # uploaded csv file. This list becomes an empty list
         # once metadata changes are written to the blank sqlite file as
         # part of the sync operation.
+        self.refresh_from_db()
```
Comment (sblack-usu) [inline hs_app_timeseries/models.py]: I don't think we should have a refresh_from_db in here.

File: hs_app_netCDF/models.py [hunk 8]
```diff
@@ -342,6 +342,7 @@ def get_hs_term_dict(self):
         return hs_term_dict
 
     def update_netcdf_file(self, user):
+        self.metadata.refresh_from_db()
```
Comment (sblack-usu) [inline hs_app_netCDF/models.py]: why did you add the refresh into here?

File: hs_core/views/utils.py [hunk 9]
```diff
@@ -544,7 +545,13 @@ def create_form(formclass, request):
 
 
 def get_my_resources_list(request):
```
Comment (sblack-usu) [inline hs_core/views/utils.py]: could you update this to take user instead of request? I don't like passing around request.

File: hs_core/models.py (hunk 6 above)
Comment (sblack-usu) [reply hs_core/models.py]: I did a quick test and got a speed up of about .5 second  on jeff's my-resource page by adding
```resource_collection = resource_collection.prefetch_related('content_type')```
to get_my_resources_list and then used content_type in place of metadata in the my-resources.html

I have no idea why we're doing this lazy loading and it doesn't seem necessary.  If you agree, we can go ahead and prefetch this data to avoid extra db calls.

File: hs_file_types/models/netcdf.py [hunk 10]
```diff
@@ -167,6 +167,7 @@ def get_html_forms(self, dataset_name_form=True, temporal_coverage=True, **kwarg
     def get_update_netcdf_file_html_form(self):
         form_action = "/hsapi/_internal/{}/update-netcdf-file/".format(self.logical_file.id)
         style = "display:none;"
+        self.refresh_from_db()
```
Comment (sblack-usu) [inline hs_file_types/models/netcdf.py]: I don't get these db refreshes.

File: hs_core/templates/pages/my-resources.html [hunk 11]
```diff
@@ -293,7 +295,7 @@ <h4 class="panel-title">
                                     Favorite
                                 {% endif %}
                             </td>
-                            <td>{{ res.updated|date:"U" }}</td>
+                            <td>{{ res.last_updated|date:"U" }}</td>
```
Comment (sblack-usu) [inline hs_core/templates/pages/my-resources.html]: Looks like I missed a last_updated, thanks for catching that.  I think it's worth making last_updated a field instead of hiding it away in metadata... helps speed things up too.  I put last_updated back to updated and got about 1 second speed up on jeff's my resource page.  I'll go ahead and create a ticket if you agree that we should add a last_updated field.  You might know something I don't.

File: hs_app_netCDF/models.py (hunk 8 above)
Comment (pkdash) [reply hs_app_netCDF/models.py]: This ensures that we get the updated value of the `metadata.is_dirty` property (which might get set due to changes to metadata elements) on which the update of netcdf file depends. This was not needed before since the call to `self.metadata` used to retrieve the metadata object from DB each time such a call was made.

File: hs_file_types/models/netcdf.py (hunk 10 above)
Comment (pkdash) [reply hs_file_types/models/netcdf.py]: For the same reason as explained previously (above).

File: hs_core/templates/pages/my-resources.html (hunk 11 above)
Comment (pkdash) [reply hs_core/templates/pages/my-resources.html]: Yes if we add the 'last_updated' to resource class we should see performance improvement in particular when we are listing many resources (like my resource page) where for each resource we are displaying last update date. When we are accessing last_updated from metadata, we are hitting the database for each resource.

File: hs_core/templates/pages/my-resources.html (hunk 2 above)
Comment (pkdash) [reply hs_core/templates/pages/my-resources.html]: Now prefetching `content_object` (metadata).

File: hs_core/views/__init__.py (hunk 3 above)
Comment (pkdash) [reply hs_core/views/__init__.py]: We no more creating this context variable.

File: hs_core/models.py (hunk 5 above)
Comment (pkdash) [reply hs_core/models.py]: This method has been removed.

File: hs_core/models.py (hunk 6 above)
Comment (pkdash) [reply hs_core/models.py]: Even with this lazy loading of 'metadata' we could have prefetched `content_object` and got the same level of performance improvement if any. However, I made the change to set the `content_object` as part of resource creation. I am also now prefetching all metadata elements (title, creators, subjects and dates). This probably helping in page loading time.

File: hs_app_timeseries/models.py (hunk 7 above)
Comment (pkdash) [reply hs_app_timeseries/models.py]: See below.

File: hs_core/views/utils.py (hunk 9 above)
Comment (pkdash) [reply hs_core/views/utils.py]: Done.

File: hs_core/templates/pages/my-resources.html (hunk 11 above)
Comment (pkdash) [reply hs_core/templates/pages/my-resources.html]: I am now prefetching 'dates' metadata element - may be we don't need to add this additional field to resource class.

File: hs_app_timeseries/tests/test_timeseries_metadata.py [hunk 12]
```diff
@@ -704,6 +707,7 @@ def test_metadata_is_dirty_flag(self):
         title = self.resTimeSeries.metadata.title
         self.resTimeSeries.metadata.update_element('title', title.id, value="New Resource Title")
         # at this point the is_dirty be set to true
+        self.resTimeSeries.metadata.refresh_from_db()
```
Comment (pkdash) [inline hs_app_timeseries/tests/test_timeseries_metadata.py]: Here is a reference to look at to undersand why `refresh_from_db()` is needed. In this case if we don't do refresh then the following statement (self.assertEqual(self.resTimeSeries.metadata.is_dirty, True) ) will fail.

File: hs_app_netCDF/models.py (hunk 8 above)
Comment (pkdash) [reply hs_app_netCDF/models.py]: Also see my comment here: https://github.com/hydroshare/hydroshare/pull/3212/files#diff-0daa45deebe2e6282237d0c4e20456c9R710

File: hs_app_timeseries/models.py (hunk 7 above)
Comment (pkdash) [reply hs_app_timeseries/models.py]: As my comment above

**Comment by pkdash** [comment]
@sblack-usu I think I have now addressed your comments/concerns.
